File: fern/snippets/providers/python/openai-responses.py
from openai import OpenAI
from composio import Composio
from composio_openai import OpenAIProvider

# Initialize Composio with OpenAI Provider
composio = Composio(provider=OpenAIProvider())
openai_client = OpenAI()

# Use a unique external ID for each of your users
user_id = "user-dev-178"

# Fetch tools for the assistant
tools = composio.tools.get(
    user_id=user_id,
    toolkits=["GITHUB", "SLACK"]
)

# Create an OpenAI assistant with Composio tools
assistant = openai_client.beta.assistants.create(
    name="Developer Assistant",
    instructions="You are a helpful assistant that can interact with GitHub and Slack.",
    model="gpt-4-turbo",
    tools=tools  # Composio tools are directly compatible
)

# Create a thread
thread = openai_client.beta.threads.create()

# Add a message to the thread
message = openai_client.beta.threads.messages.create(
    thread_id=thread.id,
    role="user",
    content="Create a GitHub issue about the new feature we discussed"
)

# Run the assistant
run = openai_client.beta.threads.runs.create(
    thread_id=thread.id,
    assistant_id=assistant.id
)

# Handle tool calls and wait for completion
run_after_tool_calls = composio.provider.wait_and_handle_assistant_tool_calls(
    user_id=user_id,
    client=openai_client,
    run=run,
    thread=thread
)

# Get the assistant's response
messages = openai_client.beta.threads.messages.list(thread_id=thread.id)
print(messages.data[0].content[0].text.value)

File: fern/snippets/providers/typescript/openairesponses.ts
import OpenAI from 'openai';
import { Composio } from '@composio/core';
import { OpenAIResponsesProvider } from '@composio/openai';

// Initialize Composio client with OpenAI Provider
const composio = new Composio({ 
    provider: new OpenAIResponsesProvider(), 
});
const openai = new OpenAI({});

// Make sure to create an auth config and a connected account for the user with gmail toolkit
// Make sure to replace "your-user-id" with the actual user ID
const userId = "your-user-id";

async function main() {
    try {
        const tools = await composio.tools.get(userId, {tools: ["GMAIL_SEND_EMAIL"]});

        const response = await openai.responses.create({
            model: "gpt-5",
            tools: tools,
            input: [
                {
                    role: "user", 
                    content: "Send an email to soham.g@composio.dev with the subject 'Running OpenAI Provider snippet' and body 'Hello from the code snippet in openai docs'"
                },
            ],
        });

        // Execute the function calls
        const result = await composio.provider.handleToolCalls(userId, response.output);
        console.log(result);
    } catch (error) {
        console.error('Error:', error);
    }
}

main();
